share/disambiguation.py

```diff
@@ -115,34 +112,15 @@ def _instance_for_node(self, node):
                 queries.append(all_query & Q(**{k: v}))
 
         if (info.all and not all_query.children) or (info.any and not queries):
-            return None
+            return []
 
         if info.matching_types:
             all_query &= Q(type__in=info.matching_types)
 
-        constrain = [Q()]
-        if hasattr(node.model, '_typedmodels_type'):
-            constrain.append(Q(type__in=node.model.get_types()))
-            constrain.append(Q(type=node.model._typedmodels_type))
-
-        for q in constrain:
-            sql, params = zip(*[concrete_model.objects.filter(all_query & query & q).query.sql_with_params() for query in queries or [Q()]])
-            found = list(concrete_model.objects.raw(' UNION '.join('({})'.format(s) for s in sql) + ' LIMIT 2;', sum(params, ())))
-
-            if not found:
-                logger.debug('No %ss found for %s %s', concrete_model, all_query & q, queries)
-                return None
-            if len(found) == 1:
-                return found[0]
-            if all_query.children:
-                logger.warning('Multiple %ss returned for %s (The main query) bailing', concrete_model, all_query)
-                break
-            if all('__' in str(query) for query in queries):
-                logger.warning('Multiple %ss returned for %s (The any query) bailing', concrete_model, queries)
-                break
-
-        logger.error('Could not disambiguate %s. Too many results found from %s %s', node.model, all_query, queries)
-        raise NotImplementedError('Multiple {0}s found'.format(node.model))
+        unmerged_query = Q(same_as__isnull=True) if hasattr(concrete_model, 'same_as') else Q()
+
+        sql, params = zip(*[concrete_model.objects.filter(unmerged_query & all_query & query).query.sql_with_params() for query in queries or [Q()]])
+        return list(concrete_model.objects.raw(' UNION '.join('({})'.format(s) for s in sql) + ';', sum(params, ())))
```

Comment:
Gooood call. I feel like the limit could be quite low, I'd be concerned about merging 10 or even 5 things together.